src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2221,114 +2202,100 @@ impl Coordinator {
         }
     }
 
+    /// Determines the query timestamp and acquires read holds on dependent sources
+    /// if necessary.
+    fn sequence_peek_timestamp(
+        &mut self,
+        session: &mut Session,
+        when: &QueryWhen,
+        cluster_id: ClusterId,
+        timeline_context: TimelineContext,
+        id_bundle: &CollectionIdBundle,
+        source_ids: &BTreeSet<GlobalId>,
+        real_time_recency_ts: Option<Timestamp>,
+    ) -> Result<TimestampDetermination<Timestamp>, AdapterError> {
+        let determination = self.determine_timestamp(
+            session,
+            id_bundle,
+            when,
+            cluster_id,
+            timeline_context.clone(),
+            real_time_recency_ts,
+        )?;
+
+        // We only track the peeks in the session if the query doesn't use AS
+        // OF or we're inside an explicit transaction. The latter case is
+        // necessary to support PG's `BEGIN` semantics, whose behavior can
+        // depend on whether or not reads have occurred in the txn.
+        if matches!(session.transaction(), &TransactionStatus::InTransaction(_))
+            || when == &QueryWhen::Immediately
```

Comment:
I know that you copied this from somewhere else, but I can't seem to figure out why this isn't `if in_immediate_multi_stmt_txn { ... }`

Reply:
Yeah, I wasn't too sure either so I left it the same. I'm thinking the intention is to track _all_ peeks within a transaction? For example:

```sql
BEGIN;

SELECT * FROM tbl AS OF AT LEAST t1;

SELECT * FROM tbl AS OF AT LEAST t2;
```

In this case, if t1 and t2 are different, adding the peek fails. Whereas we wouldn't end up attempting to track these operations if it was under `if in_immediate_multi_stmt_txn { ... }`.

Side note: Filed https://github.com/MaterializeInc/materialize/issues/19178 as this example ends up crashing the server

Reply:
Interesting. For the `SELECT ... AS OF ...` I think want to track that a peek happened, but we don't actually want to track it's timestamp. I have a sneaking suspicion that this is related to the panic you found. This is maybe more evidence that we should disallow `AS OF` in multi-statement transactions.

Either way I think this is getting out of scope for this PR. For now this is fine since it's existing behavior.